[Index Management] Enrich policies: add support for listing, executing and deleting

sabarasaba opened the pull request:
## Summary

This is the first of a series of PR's that will be merged in the [enrich-policies feature branch](https://github.com/elastic/kibana/tree/feature/enrich-policies). 

This PR creates the barebones for supporting enrich policies in index management, also gives support for listing policies, executing them and deleting them.

Note to reviewers: 
* Copy and doc links will be separately addressed with https://github.com/elastic/kibana/issues/163808
* API integration tests will be separately addressed with https://github.com/elastic/kibana/issues/163807

##### How to test
1. Start elasticsearch with `yarn es snapshot --license=trial` and serverless search solution with `yarn start`
2. Navigate to console and execute the commands listed below in order to create a policy
3. Navigate to `index management` and click on `enrich policies` tab
4. Verify that:
* The policy you just created is listed there
* Reload and search works
* Empy states (loading, error)
* Policy can be correctly executed and a toast is shown
* Policy can be correctly deleted and a toast is shown
* It shows error toast when failing to delete/execute policy (you can force a failure in the server response of the API)


<details>
<summary>Create a policy</summary>

```
PUT /users
{
  "mappings": {
    "properties": {
      "email": {
        "type": "keyword"
      },
      "first_name": {
        "type": "text"
      },
      "last_name": {
        "type": "text"
      },
      "city": {
        "type": "text"
      },
      "zip": {
        "type": "keyword"
      },
      "state": {
        "type": "keyword"
      }
    }
  }
}


PUT /_enrich/policy/my-policy
{
  "match": {
    "indices": "users",
    "match_field": "email",
    "enrich_fields": ["first_name", "last_name", "city", "zip", "state"]
  }
}
```
</details>

<details>
<summary>Screenshots empty states</summary>

// Permissions error
<img width="1507" alt="Screenshot 2023-08-16 at 13 16 41" src="https://github.com/elastic/kibana/assets/1191206/73d2364f-054c-4977-a882-91e930b72629">
// Loading state
<img width="1078" alt="Screenshot 2023-08-07 at 11 43 17" src="https://github.com/elastic/kibana/assets/1191206/6e6fc769-7c3c-4319-a1d7-15866532c71c">
// No data
<img width="1076" alt="Screenshot 2023-08-07 at 11 42 28" src="https://github.com/elastic/kibana/assets/1191206/83413bfd-2dcf-4cae-addf-6f9d479526ee">

</details>

<details>
<summary>Screenshots</summary>

<img width="1507" alt="Screenshot 2023-08-15 at 14 33 52" src="https://github.com/elastic/kibana/assets/1191206/5dba5ee7-ebe9-4634-9ab1-91333df9d801">
<img width="1509" alt="Screenshot 2023-08-15 at 14 35 33" src="https://github.com/elastic/kibana/assets/1191206/13b2f895-b3e1-44e5-aad4-b71ccda3514b">
<img width="1510" alt="Screenshot 2023-08-15 at 14 34 11" src="https://github.com/elastic/kibana/assets/1191206/8427e27d-3081-47dc-837d-9818c2ee85b7">


</details>

sabarasaba commented:
@elasticmachine merge upstream

sabarasaba commented:
@elasticmachine merge upstream

sabarasaba commented:
@elasticmachine merge upstream

elasticmachine commented:
Pinging @elastic/platform-deployment-management (Team:Deployment Management)

abdonpijpelink reviewed (approved):
Doc link change LGTM.

sabarasaba commented:
buildkite test this

yuliacech commented on x-pack/plugins/index_management/public/application/app_context.tsx:
```diff
@@ -56,6 +57,7 @@ export interface AppDependencies {
   docLinks: DocLinksStart;
   kibanaVersion: SemVer;
   theme$: Observable<CoreTheme>;
+  toasts: NotificationsSetup['toasts'];
```
We already have `services.notificationService` in the app context, I think we could just re-use it, it's like a wrapper around toasts.

yuliacech commented on x-pack/plugins/index_management/public/application/sections/home/enrich_policies_list/confirm_modals/execute_policy_modal.tsx:
```diff
@@ -0,0 +1,102 @@
+      .then(({ data, error }) => {
+        if (data) {
+          const successMessage = i18n.translate(
+            'xpack.idxMgmt.enrich_policies.executeModal.successDeleteNotificationMessage',
```
probably a typo from copying
```suggestion
            'xpack.idxMgmt.enrich_policies.executeModal.successExecuteNotificationMessage',
```

yuliacech commented on x-pack/plugins/index_management/public/application/sections/home/enrich_policies_list/confirm_modals/execute_policy_modal.tsx:
```diff
@@ -0,0 +1,102 @@
+
+        if (error) {
+          const errorMessage = i18n.translate(
+            'xpack.idxMgmt.enrich_policies.executeModal.errorDeleteNotificationMessage',
```
same here 
```suggestion
            'xpack.idxMgmt.enrich_policies.executeModal.errorExecuteNotificationMessage',
```

yuliacech commented on x-pack/plugins/index_management/public/application/sections/home/enrich_policies_list/empty_states/loading_state.tsx:
```diff
@@ -0,0 +1,28 @@
+        <EuiText color="subdued">
+          <FormattedMessage
+            id="xpack.idxMgmt.enrich_policies.list.loadingStateLabel"
+            defaultMessage="Loading index details…"
```
typo
```suggestion
            defaultMessage="Loading enrich policies…"
```

yuliacech commented on x-pack/plugins/index_management/public/application/sections/home/enrich_policies_list/policies_table/policies_table.tsx:
```diff
@@ -0,0 +1,149 @@
+
+const pagination = {
+  initialPageSize: 5,
+  pageSizeOptions: [3, 5, 8],
```
these values were probably used for testing, should we set them to something larger?

sabarasaba replied:
Nice one! 

I'll replace it with the following, which seemed to have been the agreement from the `default rows per page setting` issue
```
const pagination = {
  initialPageSize: 50,
  pageSizeOptions: [25, 50, 100],
};
```

yuliacech commented on x-pack/plugins/index_management/public/application/services/api.ts:
```diff
@@ -311,3 +312,28 @@ export function useLoadNodesPlugins() {
+
+export async function deleteEnrichPolicy(policyName: string) {
+  const result = sendRequest({
+    path: `${API_BASE_PATH}/enrich_policies/${policyName}`,
```
I think we should use the internal path for the delete and execute routes too, wdyt?

sabarasaba replied:
I think we can do it for these other two just to keep it consistent with the list enrich policies method, but I think this sort of changes we should address in a more cohesive way through the plugin as now we will have certain routes that are "internal" and others that are public. I can create an issue and perhaps we can discuss that during our next sync!

yuliacech commented on x-pack/test/functional/config.base.js:
```diff
@@ -556,7 +556,7 @@ export default async function ({ readConfigFile }) {
         index_management_user: {
           elasticsearch: {
-            cluster: ['monitor', 'manage_index_templates'],
+            cluster: ['monitor', 'manage_index_templates', 'manage_enrich'],
```
okay, I tested locally and it's not blocking IM entirely, just the enrich policies tab can't load the data.

sabarasaba replied:
Yeah it only affects `enrich policies` so the other UIs are a-ok. ES already fails with an error saying that the user doesnt have enough permissions and which one it needs. I think thats ok for now, if needed I can add a custom permission error screen in a later iteration. WDY?

<img width="1507" alt="Screenshot 2023-08-16 at 13 16 41" src="https://github.com/elastic/kibana/assets/1191206/928c8fc6-de8f-4454-a17d-df3e4a3c83f9">

yuliacech reviewed (commented):
Amazing work, @sabarasaba! 
Tested locally and everything looks great. I left a couple of minor nits in the comments. The only 2 points I'd like to get your opinion on are: if you should add the new routes under the internal path and how we should handle the permissions. 
I don't expect we should fix the permissions issue in this PR but just wanted to make sure it's planned. I think we could add a separate error message, when the permissions are missing, so that in this concrete case the users know what the error is and how to fix it. Also we need to add the new permissions to the [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-mgmt.html). This page might actually need more updates after this work is complete.

sabarasaba commented:
> Amazing work, @sabarasaba! Tested locally and everything looks great. I left a couple of minor nits in the comments. The only 2 points I'd like to get your opinion on are: if you should add the new routes under the internal path and how we should handle the permissions. I don't expect we should fix the permissions issue in this PR but just wanted to make sure it's planned. I think we could add a separate error message, when the permissions are missing, so that in this concrete case the users know what the error is and how to fix it. Also we need to add the new permissions to the [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-mgmt.html). This page might actually need more updates after this work is complete.

Thanks for the review @yuliacech! I left a comment about the permissions below your comment above.

I've addressed your comments with https://github.com/elastic/kibana/pull/163159/commits/19901fc55dde16668d902070424b917494a60209
* Marked the newly added routes as "internal" for now, although I think we need to further discuss this with the team.
* Refactored the code to consume the notificationsService directly
* Fixed the capitalization of the i18n strings
* Changed pagination size
* Few copy issues

yuliacech reviewed (approved):
Thanks a lot for addressing my comments, latest changes LGTM 🚀

sabarasaba commented:
@elasticmachine merge upstream
